File: workspace/projects/heartwood-mcp/src/tools/jt/time-entries.ts
/**
 * JT Time Entry tools — 4 tools
 */

import type { PaveClient, ToolResult } from "../../pave/index.js";
import { TIME_ENTRY_FIELDS } from "../../pave/index.js";
import type { ToolDef } from "../registry.js";

export function timeEntryTools(pave: PaveClient): ToolDef[] {
  return [
    // ── jt_create_time_entry ───────────────────────────────────────────
    {
      name: "jt_create_time_entry",
      description:
        "Create a time entry for a job. Dates must be ISO 8601 with timezone.",
      inputSchema: {
        type: "object" as const,
        properties: {
          jobId: { type: "string", description: "Job ID" },
          userId: { type: "string", description: "User ID" },
          startedAt: { type: "string", description: "Start time (ISO 8601 with timezone)" },
          endedAt: { type: "string", description: "End time (ISO 8601 with timezone)" },
          notes: { type: "string", description: "Notes (optional)" },
          costItemId: { type: "string", description: "Cost item ID (optional)" },
          type: { type: "string", description: "Time entry type (optional)" },
          isApproved: { type: "boolean", description: "Mark as approved (optional)" },
        },
        required: ["jobId", "userId", "startedAt", "endedAt"],
      },
      handler: async (params: Record<string, unknown>): Promise<ToolResult> => {
        const data: Record<string, unknown> = {
          jobId: params.jobId,
          userId: params.userId,
          startedAt: params.startedAt,
          endedAt: params.endedAt,
        };
        for (const field of ["notes", "costItemId", "type", "isApproved"]) {
          if (params[field] !== undefined) data[field] = params[field];
        }
        return pave.create("timeEntry", data, TIME_ENTRY_FIELDS);
      },
    },

    // ── jt_get_time_entries ────────────────────────────────────────────
    {
      name: "jt_get_time_entries",
      description: "Get time entries with optional filters.",
      inputSchema: {
        type: "object" as const,
        properties: {
          jobId: { type: "string", description: "Filter by job ID (optional)" },
          userId: { type: "string", description: "Filter by user ID (optional)" },
          startDate: { type: "string", description: "Filter entries after this date (optional)" },
          endDate: { type: "string", description: "Filter entries before this date (optional)" },
          isApproved: { type: "boolean", description: "Filter by approval status (optional)" },
        },
        required: [],
      },
      handler: async (params: Record<string, unknown>): Promise<ToolResult> => {
        const conditions: Array<{ field: string; operator: string; value: unknown }> = [];
        if (params.jobId) conditions.push({ field: "jobId", operator: "eq", value: params.jobId });
        if (params.userId) conditions.push({ field: "userId", operator: "eq", value: params.userId });
        if (params.startDate) conditions.push({ field: "startedAt", operator: "gte", value: params.startDate });
        if (params.endDate) conditions.push({ field: "endedAt", operator: "lte", value: params.endDate });
        if (params.isApproved !== undefined) conditions.push({ field: "isApproved", operator: "eq", value: params.isApproved });
        return pave.query({
          entity: "timeEntry",
          fields: TIME_ENTRY_FIELDS,
          filter: conditions.length > 0 ? { operator: "and", conditions } : undefined,
        });
      },
    },

    // ── jt_get_time_entry_details ──────────────────────────────────────
    {
      name: "jt_get_time_entry_details",
      description: "Get full details for a specific time entry.",
      inputSchema: {
        type: "object" as const,
        properties: {
          timeEntryId: { type: "string", description: "Time entry ID" },
        },
        required: ["timeEntryId"],
      },
      handler: async (params: Record<string, unknown>): Promise<ToolResult> => {
        return pave.read("timeEntry", params.timeEntryId as string, TIME_ENTRY_FIELDS);
      },
    },

    // ── jt_get_time_summary ────────────────────────────────────────────
    {
      name: "jt_get_time_summary",
      description: "Get aggregated time summary for a date range.",
      inputSchema: {
        type: "object" as const,
        properties: {
          startDate: { type: "string", description: "Start date (YYYY-MM-DD)" },
          endDate: { type: "string", description: "End date (YYYY-MM-DD)" },
          groupBy: { type: "string", description: "Group by field (optional)" },
          jobId: { type: "string", description: "Filter by job ID (optional)" },
          userId: { type: "string", description: "Filter by user ID (optional)" },
        },
        required: ["startDate", "endDate"],
      },
      handler: async (params: Record<string, unknown>): Promise<ToolResult> => {
        const data: Record<string, unknown> = {
          startDate: params.startDate,
          endDate: params.endDate,
        };
        if (params.groupBy) data.groupBy = params.groupBy;
        if (params.jobId) data.jobId = params.jobId;
        if (params.userId) data.userId = params.userId;
        return pave.execute({
          action: "query",
          entity: "timeSummary",
          data,
        });
      },
    },
  ];
}

File: heartwood-mcp/tools/time_entries.py
"""Time entry tools — log and query time entries."""

from __future__ import annotations

from datetime import datetime

from app import mcp
from constants import ORG_ID, DEFAULT_TIMEZONE
from pave import call_pave, flatten_nodes


@mcp.tool()
async def jt_create_time_entry(
    job_id: str,
    user_id: str,
    hours: float,
    date: str | None = None,
    description: str | None = None,
    cost_code_id: str | None = None,
    timezone: str = DEFAULT_TIMEZONE,
) -> dict:
    """Log a time entry on a job.

    Args:
        job_id: The job ID.
        user_id: The user who performed the work.
        hours: Number of hours worked.
        date: Date of the work (YYYY-MM-DD). Defaults to today.
        description: Description of work performed.
        cost_code_id: Optional cost code ID for categorization.
        timezone: Timezone for the entry (defaults to America/Denver).
    """
    params: dict = {
        "organizationId": ORG_ID,
        "jobId": job_id,
        "userId": user_id,
        "hours": hours,
        "timezone": timezone,
    }
    if date:
        params["date"] = date
    else:
        params["date"] = datetime.now().strftime("%Y-%m-%d")
    if description:
        params["description"] = description
    if cost_code_id:
        params["costCodeId"] = cost_code_id

    return await call_pave(
        "createTimeEntry",
        params=params,
        fields={"id": {}, "hours": {}, "date": {}},
    )


@mcp.tool()
async def jt_get_time_entries(
    job_id: str | None = None,
    user_id: str | None = None,
    date_from: str | None = None,
    date_to: str | None = None,
    first: int = 100,
) -> list[dict]:
    """List time entries, optionally filtered by job, user, or date range.

    Args:
        job_id: Filter by job.
        user_id: Filter by user.
        date_from: Start date (YYYY-MM-DD) — entries on or after.
        date_to: End date (YYYY-MM-DD) — entries on or before.
        first: Max results.
    """
    filter_obj: dict = {}
    if job_id:
        filter_obj["jobId"] = {"eq": job_id}
    if user_id:
        filter_obj["userId"] = {"eq": user_id}
    if date_from:
        filter_obj["date"] = {"gte": date_from}
    if date_to:
        if "date" in filter_obj:
            filter_obj["date"]["lte"] = date_to
        else:
            filter_obj["date"] = {"lte": date_to}

    params: dict = {"first": first}
    if filter_obj:
        params["filter"] = filter_obj

    result = await call_pave(
        "timeEntries",
        params=params,
        fields={
            "nodes": {
                "id": {},
                "hours": {},
                "date": {},
                "description": {},
                "userId": {},
                "jobId": {},
                "costCodeId": {},
            }
        },
    )
    return flatten_nodes(result)


@mcp.tool()
async def jt_get_time_entry_details(time_entry_id: str) -> dict:
    """Get full details for a specific time entry.

    Args:
        time_entry_id: The time entry ID.
    """
    result = await call_pave(
        "timeEntries",
        params={"filter": {"id": {"eq": time_entry_id}}, "first": 1},
        fields={
            "nodes": {
                "id": {},
                "hours": {},
                "date": {},
                "description": {},
                "userId": {},
                "user": {"id": {}, "firstName": {}, "lastName": {}},
                "jobId": {},
                "job": {"id": {}, "name": {}, "number": {}},
                "costCodeId": {},
                "createdAt": {},
            }
        },
    )
    nodes = flatten_nodes(result)
    return nodes[0] if nodes else {}


@mcp.tool()
async def jt_update_time_entry(
    time_entry_id: str,
    hours: float | None = None,
    date: str | None = None,
    description: str | None = None,
) -> dict:
    """Update a time entry.

    Args:
        time_entry_id: The time entry ID.
        hours: New hours value.
        date: New date (YYYY-MM-DD).
        description: New description.
    """
    params: dict = {"id": time_entry_id}
    if hours is not None:
        params["hours"] = hours
    if date is not None:
        params["date"] = date
    if description is not None:
        params["description"] = description

    return await call_pave(
        "updateTimeEntry",
        params=params,
        fields={"id": {}, "hours": {}, "date": {}, "description": {}},
    )


@mcp.tool()
async def jt_delete_time_entry(time_entry_id: str) -> dict:
    """Delete a time entry.

    Args:
        time_entry_id: The time entry ID to delete.
    """
    return await call_pave(
        "deleteTimeEntry",
        params={"id": time_entry_id},
        fields={"id": {}},
    )


@mcp.tool()
async def jt_get_time_summary(
    job_id: str | None = None,
    user_id: str | None = None,
    date_from: str | None = None,
    date_to: str | None = None,
) -> dict:
    """Get a summary of time entries (total hours) with optional filters.

    Returns total hours and entry count rather than individual entries.

    Args:
        job_id: Filter by job.
        user_id: Filter by user.
        date_from: Start date (YYYY-MM-DD).
        date_to: End date (YYYY-MM-DD).
    """
    entries = await jt_get_time_entries(
        job_id=job_id,
        user_id=user_id,
        date_from=date_from,
        date_to=date_to,
        first=500,
    )
    total_hours = sum(e.get("hours", 0) for e in entries)
    return {
        "totalHours": total_hours,
        "entryCount": len(entries),
        "entries": entries,
    }
